tests/test_forwarder.py

```diff
@@ -292,12 +338,247 @@ def _test_server(self, server):
+                local_bind_addresses=[('127.0.0.1', self.eport),
+                                      ('127.0.0.1', self.randomize_eport())]
+            )
+
+    def test_remotebindaddress_and_remotebindaddresses_are_exclusive(self):
+        """
+        Test that arguments remote_bind_address and remote_bind_addresses
+        cannot be used together
+        """
+        with self.assertRaises(ValueError):
+            SSHTunnelForwarder(
+                (self.saddr, self.sport),
+                ssh_username=SSH_USERNAME,
+                ssh_password=SSH_PASSWORD,
+                remote_bind_address=(self.eaddr, self.eport),
+                remote_bind_addresses=[(self.eaddr, self.eport),
+                                       (self.eaddr, self.randomize_eport())]
+            )
+
+    def test_no_remote_bind_address_raises_exception(self):
+        """
+        When no remote_bind_address or remote_bind_addresses are specified, a
+        ValueError exception should be raised
+        """
+        with self.assertRaises(ValueError):
+            SSHTunnelForwarder(
+                (self.saddr, self.sport),
+                ssh_username=SSH_USERNAME,
+                use_ssh_config=False
+            )
+
+    def test_reading_from_a_bad_sshconfigfile_does_not_raise_error(self):
+        """
+        Test that when a bad ssh_config file is found, a warning is shown
+        but no exception is raised
+        """
+        ssh_config_file = 'not_existing_file'
+
+        SSHTunnelForwarder(
+            (self.saddr, self.sport),
+            ssh_username=SSH_USERNAME,
+            ssh_password=SSH_PASSWORD,
+            remote_bind_address=(self.eaddr, self.eport),
+            local_bind_address=('127.0.0.1', self.randomize_eport()),
+            ssh_config_file=ssh_config_file
+        )
+        logged_message = 'Could not read SSH configuration file: {}'.format(
+                             ssh_config_file
+                         )
+        self.assertIn(logged_message, self.sshtunnel_log_messages['warning'])
+
+    def test_not_setting_password_or_pkey_raises_error(self):
+        """
+        Test that when a no authentication method is specified, an exception is
+        raised
+        """
+        with self.assertRaises(ValueError):
+            SSHTunnelForwarder(
+                (self.saddr, self.sport),
+                ssh_username=SSH_USERNAME,
+                remote_bind_address=(self.eaddr, self.eport),
+                use_ssh_config=False
+            )
+
+    def test_deprecate_warnings_are_shown(self):
+        """Test that when using deprecate arguments a warning is logged"""
+        warnings.simplefilter("always")  # don't ignore DeprecationWarnings
+        with warnings.catch_warnings(record=True) as w:
+            for deprecated_arg in ['ssh_address', 'ssh_host']:
+                eval(
+                    SSHTunnelForwarder.__name__ + '({}={})'.format(
```

Comment:
dirty hack :) Please use the different `kwargs` objects for `SSHTunnelForwarder(**kwargs)`. Like: `kwargs = dict(ssh_username=SSH_USERNAME, ..); kwargs.update(dict(..))`